README.md
# systemd-timers

[![Build Status](https://travis-ci.org/dtan4/systemd-timers.svg?branch=master)](https://travis-ci.org/dtan4/systemd-timers)

Better `systemctl list-timers`

![example](_images/example.png)

## VS. `systemctl list-timers`

systemd has `systemctl list-timers` command to list installed systemd timers.

```bash
core@ip-10-1-2-255 ~ $ sudo systemctl list-timers
NEXT                         LEFT         LAST                         PASSED       UNIT                           ACTIVATES
Fri 2017-02-03 13:54:00 JST  12s left     Fri 2017-02-03 13:53:02 JST  44s ago      failed.timer                   failed.service
Fri 2017-02-03 19:03:00 JST  5h 9min left Fri 2017-02-03 05:03:18 JST  8h ago       docker-cleanup.timer           docker-cleanup.service
Fri 2017-02-03 21:51:07 JST  7h left      Thu 2017-02-02 21:51:07 JST  16h ago      systemd-tmpfiles-clean.timer   systemd-tmpfiles-clean.service
```

We cannot check task execution result (whether last execution succeeded or failed) at this view.
We have to do `systemctl status <unit>` for each task.
It's so bothersome (at least for me).

This `systemd-timers` show task execution result at once. They are colorized; success is green, failed is red.

```bash
core@ip-10-1-2-255 ~ $ systemd-timers
UNIT                      LAST                                          RESULT   EXECUTION TIME          NEXT                           SCHEDULE
update-valec              2017-02-03 03:00:17 +0900 JST (15 hours ago)  success  7 seconds               2017-02-04 03:00:00 +0900 JST  *-*-* 03:00:00
backup-prod               2017-02-03 06:10:17 +0900 JST (12 hours ago)  failed   Less than a second      2017-02-04 06:10:00 +0900 JST  *-*-* 06:10:00
backup-stag               2017-02-03 06:05:08 +0900 JST (12 hours ago)  failed   Less than a second      2017-02-04 06:05:00 +0900 JST  *-*-* 06:05:00
```


CHANGELOG.md
# [v0.1.0](https://github.com/dtan4/systemd-timers/releases/tag/v0.1.0) (2017-02-03)

Initial release.


table_test.go
package main

import (
	"testing"
)

func TestFormatExecutionTime(t *testing.T) {
	testcases := []struct {
		executionTime uint64
		expected      string
	}{
		{
			executionTime: 0,
			expected:      "n/a",
		},
		{
			executionTime: 100,
			expected:      "Less than a second",
		},
		{
			executionTime: 1 * 1000 * 1000,
			expected:      "1 second",
		},
		{
			executionTime: 20 * 1000 * 1000,
			expected:      "20 seconds",
		},
		{
			executionTime: 60 * 1000 * 1000,
			expected:      "1 minute",
		},
		{
			executionTime: 61 * 1000 * 1000,
			expected:      "1 minute 1 second",
		},
		{
			executionTime: 62 * 1000 * 1000,
			expected:      "1 minute 2 seconds",
		},
		{
			executionTime: 120 * 1000 * 1000,
			expected:      "2 minutes",
		},
		{
			executionTime: 121 * 1000 * 1000,
			expected:      "2 minutes 1 second",
		},
		{
			executionTime: 142 * 1000 * 1000,
			expected:      "2 minutes 22 seconds",
		},
		{
			executionTime: 192 * 60 * 1000 * 1000,
			expected:      "192 minutes",
		},
	}

	for _, testcase := range testcases {
		got := formatExecutionTime(testcase.executionTime)
		if got != testcase.expected {
			t.Errorf("got: %q, expected: %q", got, testcase.expected)
		}
	}
}


systemd/daemon.go
package systemd

import (
	"strings"
	"time"

	"github.com/coreos/go-systemd/dbus"
)

const (
	serviceSuffix   = ".service"
	serviceUnitType = "Service"
	timerSuffix     = ".timer"
	timerUnitType   = "Timer"
)

// Client represents systemd D-Bus API client.
type Client struct {
	conn *dbus.Conn
}

// NewClient creates new Client object
func NewClient(conn *dbus.Conn) *Client {
	return &Client{
		conn: conn,
	}
}

// NewConn establishes a new connection to D-Bus
func NewConn() (*dbus.Conn, error) {
	return dbus.New()
}

// ListTimers returns installed systemd timers
func (c *Client) ListTimers() ([]*Timer, error) {
	units, err := c.conn.ListUnits()
	if err != nil {
		return []*Timer{}, err
	}

	timers := []*Timer{}

	for _, unit := range units {
		if !strings.HasSuffix(unit.Name, timerSuffix) {
			continue
		}

		timer := NewTimer(strings.TrimSuffix(unit.Name, timerSuffix))

		serviceProps, err := c.conn.GetUnitTypeProperties(timer.ServiceName(), serviceUnitType)
		if err != nil {
			return []*Timer{}, err
		}

		var execMainStartTimestamp, execMainExitTimestamp uint64

		if v, ok := serviceProps["ExecMainStartTimestamp"]; ok {
			if v2, ok2 := v.(uint64); ok2 {
				execMainStartTimestamp = v2
			}
		}

		if v, ok := serviceProps["ExecMainExitTimestamp"]; ok {
			if v2, ok2 := v.(uint64); ok2 {
				execMainExitTimestamp = v2
			}
		}

		if execMainStartTimestamp < execMainExitTimestamp {
			timer.LastExecutionTime = execMainExitTimestamp - execMainStartTimestamp
		} else {
			timer.LastExecutionTime = 0
		}

		if v, ok := serviceProps["Result"]; ok {
			if result, ok2 := v.(string); ok2 {
				timer.Result = result
			}
		}

		timerProps, err := c.conn.GetUnitTypeProperties(timer.TimerName(), timerUnitType)
		if err != nil {
			return []*Timer{}, err
		}

		if v, ok := timerProps["TimersCalendar"]; ok {
			if s, ok2 := v.([][]interface{}); ok2 {
				// []interface {}{"OnCalendar", "*-*-* 06,18:00:00", 0x5475da471b800}
				if len(s) > 0 && len(s[0]) > 1 {
					if schedule, ok3 := s[0][1].(string); ok3 {
						timer.Schedule = schedule
					}
				}
			}
		}

		if v, ok := timerProps["LastTriggerUSec"]; ok {
			if lastTriggerUSec, ok2 := v.(uint64); ok2 {
				if lastTriggerUSec == 0 {
					timer.LastTriggered = time.Time{}
				} else {
					timer.LastTriggered = time.Unix(int64(lastTriggerUSec)/1000/1000, 0)
				}
			}
		}

		if v, ok := timerProps["NextElapseUSecRealtime"]; ok {
			if nextElapseUSecRealtime, ok2 := v.(uint64); ok2 {
				if nextElapseUSecRealtime == 0 {
					timer.NextElapse = time.Time{}
				} else {
					nextElapse := time.Unix(int64(nextElapseUSecRealtime)/1000/1000, 0)

					if nextElapse.After(time.Now()) {
						timer.NextElapse = nextElapse
					} else {
						timer.NextElapse = time.Time{}
					}
				}
			}
		}

		timers = append(timers, timer)
	}

	return timers, nil
}

// Timer represents systemd timer
type Timer struct {
	Name              string    `json:"name"`
	Schedule          string    `json:"schedule"`
	LastTriggered     time.Time `json:"last_triggered"`
	NextElapse        time.Time `json:"next_elapse"`
	Result            string    `json:"result"`
	Active            bool      `json:"active"`
	LastExecutionTime uint64    `json:"last_execution_time"`
}

// NewTimer creates new Timer object
func NewTimer(name string) *Timer {
	return &Timer{
		Name: name,
	}
}

// ServiceName returns service unit name
func (t *Timer) ServiceName() string {
	return t.Name + serviceSuffix
}

// TimerName returns timer unit name
func (t *Timer) TimerName() string {
	return t.Name + timerSuffix
}


table.go
package main

import (
	"bytes"
	"fmt"
	"strconv"
	"strings"
	"text/tabwriter"

	"github.com/dtan4/systemd-timers/systemd"
	"github.com/dustin/go-humanize"
	"github.com/reconquest/loreley"
)

const (
	oneSecond = 1000 * 1000
	oneMinute = 60 * oneSecond
)

var (
	headers = []string{
		"UNIT",
		"LAST",
		"RESULT",
		"EXECUTION TIME",
		"NEXT",
		"SCHEDULE",
	}
)

func generateTable(timers []*systemd.Timer) (string, error) {
	buf := &bytes.Buffer{}

	w := tabwriter.NewWriter(buf, 0, 0, 2, ' ', tabwriter.FilterHTML)
	fmt.Fprintln(w, strings.Join(headers, "\t"))

	for _, timer := range timers {
		var lastTriggered, result, nextElapse string

		if timer.LastTriggered.IsZero() {
			lastTriggered = "n/a"
			result = "n/a"
		} else {
			lastTriggered = fmt.Sprintf("%s (%s)", timer.LastTriggered.Local().String(), humanize.Time(timer.LastTriggered))

			if timer.Result == "success" {
				result = "<fg 2>success<reset>"
			} else {
				result = "<fg 1>failed<reset>"
			}
		}

		if timer.NextElapse.IsZero() {
			nextElapse = "n/a"
		} else {
			nextElapse = timer.NextElapse.Local().String()
		}

		fmt.Fprintln(w, strings.Join([]string{
			timer.Name,
			lastTriggered,
			result,
			formatExecutionTime(timer.LastExecutionTime),
			nextElapse,
			timer.Schedule,
		}, "\t"))
	}

	w.Flush()

	loreley.DelimLeft = "<"
	loreley.DelimRight = ">"

	table, err := loreley.CompileAndExecuteToString(
		buf.String(),
		nil,
		nil,
	)
	if err != nil {
		return "", err
	}

	return table, nil
}

func formatExecutionTime(executionTime uint64) string {
	if executionTime == 0 {
		return "n/a"
	}

	if executionTime < oneSecond {
		return "Less than a second"
	}

	if executionTime < 2*oneSecond {
		return "1 second"
	}

	if executionTime < 60*oneSecond {
		return fmt.Sprintf("%s seconds", strconv.Itoa(int(executionTime/oneSecond)))
	}

	if executionTime < 61*oneSecond {
		return "1 minute"
	}

	if executionTime < 62*oneSecond {
		return "1 minute 1 second"
	}

	if executionTime < 2*oneMinute {
		return fmt.Sprintf("1 minute %s seconds", strconv.Itoa(int((executionTime-oneMinute)/oneSecond)))
	}

	if (executionTime-oneMinute)/oneSecond%60 < 1 {
		return fmt.Sprintf("%s minutes", strconv.Itoa(int(executionTime/oneSecond/60)))
	}

	if (executionTime-oneMinute)/oneSecond%60 < 2 {
		return fmt.Sprintf("%s minutes 1 second", strconv.Itoa(int(executionTime/oneSecond/60)))
	}

	return fmt.Sprintf("%s minutes %s seconds", strconv.Itoa(int(executionTime/oneSecond/60)), strconv.Itoa(int((executionTime-oneMinute)/oneSecond%60)))
}
